# Unit 12 - Tales from the Crypto

---


## 1. Sentiment Analysis

Use the [newsapi](https://newsapi.org/) to pull the latest news articles for Bitcoin and Ethereum and create a DataFrame of sentiment scores for each coin.

Use descriptive statistics to answer the following questions:
1. Which coin had the highest mean positive score?
2. Which coin had the highest negative score?
3. Which coin had the highest positive score?

In [1]:
# Initial imports
import os
import pandas as pd
from dotenv import load_dotenv
import nltk as nltk
from newsapi.newsapi_client import NewsApiClient
nltk.download('vader_lexicon')
from nltk.sentiment.vader import SentimentIntensityAnalyzer
analyzer = SentimentIntensityAnalyzer()

%matplotlib inline

[nltk_data] Downloading package vader_lexicon to
[nltk_data]     C:\Users\Michael\AppData\Roaming\nltk_data...
[nltk_data]   Package vader_lexicon is already up-to-date!


In [2]:
# Read your api key environment variable
load_dotenv()
api_key = os.getenv("NEWS_API_KEY")

In [3]:
type(api_key)

str

In [4]:
# Create a newsapi client
newsapi = NewsApiClient(api_key = api_key)

In [5]:
# Fetch the Bitcoin news articles
btc_articles = newsapi.get_everything(q='bitcoin', language="en", sort_by="relevancy")
btc_articles

{'status': 'ok',
 'totalResults': 6243,
 'articles': [{'source': {'id': None, 'name': 'Blogspot.com'},
   'author': 'noreply@blogger.com (Unknown)',
   'title': "Elon Musk reveals who bitcoin's creator Satoshi Nakamoto might be",
   'description': 'Musk.MARK RALSTON/AFP via Getty Images\r\nElon Musk seems to agree with many that hyper-secret cryptocurrency expert Nick Szabo could be Satoshi Nakamoto, the mysterious creator of the digital currency Bitcoin.\xa0"You can watch ideas evolve before Bitcoin was lau…',
   'url': 'https://techncruncher.blogspot.com/2021/12/elon-musk-reveals-who-bitcoins-creator.html',
   'urlToImage': 'https://blogger.googleusercontent.com/img/a/AVvXsEik_48hPzMzsDzwfdUeHj4jNGqYGevEuVjTTPkAGTu9bRN3oePxV6bxF897GK8Az3AaSqUOalYXNG4HSCy0fW5KUHuruCWP8hAfZxgrgbzh-dsbLM9jxyFGCthOZdBCa1dNkqk6mrVl0VtflsV2VvKXfGnwL6-68m-mxp7qHJuLlvqGIahZ9YDe5mt97w=w1200-h630-p-k-no-nu',
   'publishedAt': '2021-12-29T20:41:00Z',
   'content': 'Musk.MARK RALSTON/AFP via Getty Images\r\nElon

In [6]:
# Fetch the Ethereum news articles
eth_articles = newsapi.get_everything(q='etherum', language="en", sort_by="relevancy")
eth_articles

{'status': 'ok',
 'totalResults': 17,
 'articles': [{'source': {'id': None, 'name': 'Motley Fool'},
   'author': 'newsfeedback@fool.com (Rick Munarriz)',
   'title': 'Ethereum: Buy the Dip?',
   'description': 'The popular crypto has had a rough start to 2022. You can be opportunistic, or you can head to the sidelines.',
   'url': 'https://www.fool.com/investing/2022/01/06/ethereum-buy-the-dip/',
   'urlToImage': 'https://g.foolcdn.com/editorial/images/660176/gettyimages-1317424267.jpg',
   'publishedAt': '2022-01-06T14:35:00Z',
   'content': "It wasn't just stocks taking a hit on Wednesday. Cryptocurrencies also corrected sharply as the market swooned, and Ethereum(CRYPTO:ETH) wasn't immune to the drawdown. The world's second-most-valuabl… [+3427 chars]"},
  {'source': {'id': None, 'name': 'Fox Business'},
   'author': 'Peter Aitken',
   'title': 'RadioShack to rebrand as cryptocurrency exchange platform - Fox Business',
   'description': 'Retail Ecommerce Ventures acquired RadioShack

In [7]:
# Create the Bitcoin sentiment scores DataFrame
btc_sentiments = []

for article in btc_articles["articles"]:
    try:
        text = article["content"]
        date = article["publishedAt"][:10]
        sentiment = analyzer.polarity_scores(text)
        compound = sentiment["compound"]
        pos = sentiment["pos"]
        neu = sentiment["neu"]
        neg = sentiment["neg"]
        
        btc_sentiments.append({
            "text": text,
            "date": date,
            "compound": compound,
            "positive": pos,
            "negative": neg,
            "neutral": neu
            
        })
    
    except AttributeError:
        pass

btc_df = pd.DataFrame(btc_sentiments)

cols = ["date", "text", "compound", "positive", "negative", "neutral"]
btc_df = btc_df[cols]
btc_df

date                                               text  compound  \
0   2021-12-29  Musk.MARK RALSTON/AFP via Getty Images\r\nElon...    0.3612   
1   2021-12-12  Image source, AFP via Getty Images\r\nImage ca...   -0.4019   
2   2022-01-11  A cryptocurrency CEO has become one of the ric...    0.6369   
3   2021-12-12  People aren't done hijacking major politicians...    0.5267   
4   2021-12-23  Dec 23 (Reuters) - The U.S. Securities and Exc...    0.5859   
5   2022-01-07  Between Bitcoin-themed superspreader events an...   -0.3716   
6   2021-12-12  Representation of cryptocurrency bitcoin is se...    0.0000   
7   2021-12-28  A representation of the virtual cryptocurrency...    0.0000   
8   2022-01-07  LONDON, Jan 7 (Reuters) - Bitcoin on Friday sl...   -0.3818   
9   2022-01-11  After one of roughest patches ever for Bitcoin...   -0.5563   
10  2022-01-10  ** Shares of cryptocurrency and blockchain-rel...    0.3818   
11  2022-01-10  A representation of the virtual cryptocurrency...    0.0000   
12  2022-01-06  In the spring of 2021, the encrypted communica...    0.0000   
13  2021-12-11  "Bitcoin mining is a process of competitively ...    0.2500   
14  2022-01-10  Welcome to Kitco News' 2022 outlook series. Th...    0.1531   
15  2022-01-06  Bitcoin slumped Wednesday and into Thursday.Ji...    0.0000   
16  2021-12-17  MEXICO CITY, Dec 17 (Reuters) - Mexico's Grupo...    0.7906   
17  2021-12-24  The Fine Print: The following comments are own...   -0.0418   
18  2021-12-16  Bitcoin's most notable feature is its decentra...    0.4019   
19  2022-01-10  I still wouldn't buy it. Cryptocurrency is a c...   -0.8622   

    positive  negative  neutral  
0      0.077     0.000    0.923  
1      0.000     0.078    0.922  
2      0.148     0.000    0.852  
3      0.112     0.000    0.888  
4      0.142     0.000    0.858  
5      0.000     0.074    0.926  
6      0.000     0.000    1.000  
7      0.000     0.000    1.000  
8      0.000     0.077    0.923  
9      0.063     0.156    0.781  
10     0.128     0.057    0.815  
11     0.000     0.000    1.000  
12     0.000     0.000    1.000  
13     0.129     0.074    0.797  
14     0.076     0.061    0.863  
15     0.000     0.000    1.000  
16     0.211     0.000    0.789  
17     0.079     0.086    0.835  
18     0.117     0.000    0.883  
19     0.069     0.266    0.665

In [8]:
# Create the Ethereum sentiment scores DataFrame
eth_sentiments = []

for article in eth_articles["articles"]:
    try:
        text = article["content"]
        date = article["publishedAt"][:]
        sentiment = analyzer.polarity_scores(text)
        compound = sentiment["compound"]
        pos = sentiment["pos"]
        neu = sentiment["neu"]
        neg = sentiment["neg"]
        
        eth_sentiments.append({
            "text": text,
            "date": date,
            "compound": compound,
            "positive": pos,
            "negative": neg,
            "neutral": neu
            
        })
    
    except AttributeError:
        pass

eth_df = pd.DataFrame(eth_sentiments)

cols = ["date", "text", "compound", "positive", "negative", "neutral"]
eth_df = eth_df[cols]
eth_df.head()

date                                               text  \
0  2022-01-06T14:35:00Z  It wasn't just stocks taking a hit on Wednesda...   
1  2021-12-31T15:26:01Z  RadioShack announced plans to rebrand as a cry...   
2  2021-12-27T23:28:58Z  Crypto has been many things in its short histo...   
3  2022-01-06T18:49:00Z  The Blockchain-based Carbon Credit Ecosystem a...   
4  2021-12-12T21:50:50Z  Only 1% of NFT assets are sold for over $1,500...   

   compound  positive  negative  neutral  
0   -0.2235     0.000     0.065    0.935  
1    0.3818     0.082     0.000    0.918  
2    0.0000     0.000     0.000    1.000  
3    0.3818     0.088     0.000    0.912  
4    0.1779     0.046     0.000    0.954

In [9]:
# Describe the Bitcoin Sentiment
btc_df.describe()

compound   positive   negative    neutral
count  20.000000  20.000000  20.000000  20.000000
mean    0.073625   0.067550   0.046450   0.886000
std     0.433265   0.065399   0.068545   0.091172
min    -0.862200   0.000000   0.000000   0.665000
25%    -0.124250   0.000000   0.000000   0.830000
50%     0.000000   0.072500   0.000000   0.885500
75%     0.386825   0.119750   0.074750   0.944500
max     0.790600   0.211000   0.266000   1.000000

In [10]:
# Describe the Ethereum Sentiment
eth_df.describe()

compound   positive   negative    neutral
count  17.000000  17.000000  17.000000  17.000000
mean    0.250335   0.077882   0.015353   0.906706
std     0.349114   0.075940   0.029242   0.078226
min    -0.223500   0.000000   0.000000   0.791000
25%     0.000000   0.000000   0.000000   0.822000
50%     0.226300   0.070000   0.000000   0.930000
75%     0.381800   0.124000   0.000000   0.954000
max     0.888100   0.209000   0.082000   1.000000

### Questions:

Q: Which coin had the highest mean positive score?

A: Etherum had the highest mean positive score of 0.081389

Q: Which coin had the highest compound score?

A: Etherum had the highest compound score of 0.888100

Q. Which coin had the highest positive score?

A: Bitcoin had the highest positive score of 0.211

---

## 2. Natural Language Processing
---
###   Tokenizer

In this section, you will use NLTK and Python to tokenize the text for each coin. Be sure to:
1. Lowercase each word.
2. Remove Punctuation.
3. Remove Stopwords.

In [11]:
from nltk.tokenize import word_tokenize, sent_tokenize
from nltk.corpus import stopwords
from nltk.stem import WordNetLemmatizer, PorterStemmer
from string import punctuation
import re

In [12]:
# Instantiate the lemmatizer
lemmatizer = WordNetLemmatizer()

# Create a list of stopwords
sw = set(stopwords.words('english'))

# Expand the default stopwords list if necessary


In [34]:
# Complete the tokenizer function
def tokenizer(text): 
    
    """Tokenizes text."""
    
    # Remove the punctuation from text
    regex = re.compile("[^a-zA-Z ]")
    re_clean = regex.sub('', text)
   
    # Create a tokenized list of the words
    words = word_tokenize(re_clean)    
    
    # Lemmatize words into root words
    lem = [lemmatizer.lemmatize(word) for word in words]
   
    # Convert the words to lowercase
    words.str.lower()
    
    # Remove the stop words
    output = [word.lower() for word in lem if word.lower() not in sw]
    
    return tokens

In [35]:
# Create a new tokens column for Bitcoin
btc_df["tokens"] = btc_df.text.apply(tokenizer)
btc_df.head()

AttributeError: 'list' object has no attribute 'str'

In [ ]:
# Create a new tokens column for Ethereum
for word in word_tokenized:
    words = []
    for sent in eth_df(text):
        words = words + word_tokenize(sent)
    
    word_tokenized.append(words)
    
eth_df_tokenized = pd.DataFrame({'date':date,
                                'text': text,
                                'compound': compound,
                                'positive': positive,
                                'negative': negative,
                                'neutral': neutral,
                                'word_tokenized': word_tokenized
                                })
eth_df_tokenized = ids

eth_df_tokenized.head()

---

### NGrams and Frequency Analysis

In this section you will look at the ngrams and word frequency for each coin. 

1. Use NLTK to produce the n-grams for N = 2. 
2. List the top 10 words for each coin. 

In [29]:
from collections import Counter
from nltk import ngrams

In [ ]:
# Generate the Bitcoin N-grams where N=2
btc_processed = 

In [ ]:
# Generate the Ethereum N-grams where N=2
# YOUR CODE HERE!

In [ ]:
# Function token_count generates the top 10 words for a given coin
def token_count(tokens, N=3):
    """Returns the top N tokens from the frequency count"""
    return Counter(tokens).most_common(N)

In [ ]:
# Use token_count to get the top 10 words for Bitcoin
# YOUR CODE HERE!

In [ ]:
# Use token_count to get the top 10 words for Ethereum
# YOUR CODE HERE!

---

### Word Clouds

In this section, you will generate word clouds for each coin to summarize the news for each coin

In [ ]:
from wordcloud import WordCloud
import matplotlib.pyplot as plt
plt.style.use('seaborn-whitegrid')
import matplotlib as mpl
mpl.rcParams['figure.figsize'] = [20.0, 10.0]

In [ ]:
# Generate the Bitcoin word cloud
# YOUR CODE HERE!

In [ ]:
# Generate the Ethereum word cloud
# YOUR CODE HERE!

---
## 3. Named Entity Recognition

In this section, you will build a named entity recognition model for both Bitcoin and Ethereum, then visualize the tags using SpaCy.

In [ ]:
import spacy
from spacy import displacy

In [ ]:
# Download the language model for SpaCy
# !python -m spacy download en_core_web_sm

In [ ]:
# Load the spaCy model
nlp = spacy.load('en_core_web_sm')

---
### Bitcoin NER

In [ ]:
# Concatenate all of the Bitcoin text together
# YOUR CODE HERE!

In [ ]:
# Run the NER processor on all of the text
# YOUR CODE HERE!

# Add a title to the document
# YOUR CODE HERE!

In [ ]:
# Render the visualization
# YOUR CODE HERE!

In [ ]:
# List all Entities
# YOUR CODE HERE!

---

### Ethereum NER

In [ ]:
# Concatenate all of the Ethereum text together
# YOUR CODE HERE!

In [ ]:
# Run the NER processor on all of the text
# YOUR CODE HERE!

# Add a title to the document
# YOUR CODE HERE!

In [ ]:
# Render the visualization
# YOUR CODE HERE!

In [ ]:
# List all Entities
# YOUR CODE HERE!

---